go_1_16: init at 1.16

c00w opened the pull request:
This allows us to sort out go1.16 in time for its release.

Changes are minor - I ended up just patching the ssl certs at the root
file, rather than trying to keep up with the various darwin changes.

The externalnetwork test helper location changed, to so I had to update
that patch as well.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
go1.16beta1 is available - go1.16 should drop in febuary.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

zowoq commented on pkgs/development/compilers/go/1.16.nix:
```diff
@@ -0,0 +1,260 @@
+  disallowedReferences = [ goBootstrap ];
+
+  meta = with stdenv.lib; {
+    branch = "1.15";
```
```suggestion
    branch = "1.16";
```

c00w replied:
Done

zowoq commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9571,6 +9571,13 @@ in
     buildPackages = buildPackages // { stdenv = gcc8Stdenv; };
   });
 
```
Can you add `buildGo116{Module,Package}` please?

c00w replied:
Done

c00w committed:
go_1_16: init at 1.16

Changes are minor - I ended up just patching the ssl certs at the root
file, rather than trying to keep up with the various darwin changes.

The externalnetwork test helper location changed, to so I had to update
that patch as well.

- Add xcbuild as propagatedBuildInput on darwin https://github.com/golang/go/commit/7e25bdba5e79d223566745fca2410f725a6dedda

zowoq commented:
Built on darwin.

c00w commented:
@kalbasit  @Mic92  - any thoughts / chance you can merge?

zowoq commented on pkgs/development/compilers/go/1.16.nix:
```diff
@@ -0,0 +1,259 @@
+  disallowedReferences = [ goBootstrap ];
+
+  meta = with stdenv.lib; {
+    homepage = "http://golang.org/";
```
We removed `meta.branch` in https://github.com/NixOS/nixpkgs/pull/108893.

Mic92 replied:
I think this is fixed now?

zowoq replied:
Yes, just explaining why I removed it when earlier I had asked for it to be updated.

c00w replied:
I'm a strong fan of explanations :) Also fixed.

zowoq commented:
Updated to `1.16rc1` and applied two treewide changes to match `go_1_*` on master:
- `pkgconfig` -> `pkg-config`
- `stdenv.lib` -> `lib`

zowoq commented:
Updated the `GO_NO_VENDOR_CHECKS` patch for 1.16 that it's in master.

cc @adisbladis

bobrik commented:
Either this or #112248 will need a rebase once one of them is merged.

bobrik commented on pkgs/development/compilers/go/1.16.nix:
```diff
@@ -0,0 +1,260 @@
+    ++ optionals stdenv.isLinux [ stdenv.cc.libc.out ]
+    ++ optionals (stdenv.hostPlatform.libc == "glibc") [ stdenv.cc.libc.static ];
+
+  depsTargetTargetPropagated = optionals stdenv.isDarwin [ Security Foundation ];
```
I think you also need the following for 1.16 (plus `xcbuild` in inputs):

```
propagatedBuildInputs = optionals stdenv.isDarwin [ xcbuild ];
```

See: https://github.com/golang/go/commit/7e25bdba5e79d223566745fca2410f725a6dedda

Otherwise you won't be able to link packages inside of Nix build environment:

```
/nix/store/61kpq6gcyp94j87ni6i12vl4vpmsgb02-go-1.16rc1/share/go/pkg/tool/darwin_amd64/link: /nix/store/61kpq6gcyp94j87ni6i12vl4vpmsgb02-go-1.16rc1/share/go/pkg/tool/darwin_amd64/link: running dsymutil failed: exec: "xcrun": executable file not found in $PATH
```

Go will still work outside of Nix if you run `go`, because it will find `/usr/bin/xcrun` from developer tools.

kalbasit commented:
Please update to reflect the removal of go_bootstrap in #113113

bobrik commented:
Changes from #112248 should also be picked up.

zowoq commented:
Updated/rebased on master to include https://github.com/NixOS/nixpkgs/pull/112248 and https://github.com/NixOS/nixpkgs/pull/113113.

bobrik commented:
Good to merge now? I can address https://github.com/NixOS/nixpkgs/pull/107937#discussion_r571517437 in a separate PR.

zowoq commented:
> Good to merge now?

I was going to wait until 1.16 is released and merge that.

bobrik commented:
Seems reasonable to incorporate my change to not break darwin down to road then.

bobrik commented:
Today is the day: https://github.com/golang/go/releases/tag/go1.16.

The source file is already available at the expected URL, even if https://golang.org/dl/ is not yet updated.

zowoq commented:
> ... even if https://golang.org/dl/ is not yet updated.

It's been updated.

> Seems reasonable to incorporate my change ...

Added `xcbuild`.

bobrik commented:
LGTM.

zowoq commented:
Built on `x86_64-{darwin,linux}`
